### PR TITLE
Remove Iceberg materialized view storage tables from metastores

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Storage tables clutter the namespace, and do not need a separate metastore entry. The pointer to the current metadata file is instead stored in the materialized view properties.

This applies only to newly created materialized views, existing ones are not effected.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Newly created Materialized Views will not create a separate entry in the metastore for a storage table.
```

### CODE REVIEW

**Review by findepi — COMMENTED** [review]
(skimming)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 1]
```diff
@@ -221,6 +228,67 @@ protected void deleteTableDirectory(TrinoFileSystem fileSystem, SchemaTableName
         }
     }
 
+    protected String createMaterializedViewStorage(ConnectorSession session, SchemaTableName viewName, ConnectorMaterializedViewDefinition definition)
```
[the remark this answers is not in the document]
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: I left the existing method in just in case we need to disable this new code, but yes.

[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: is the existing method now unused?

[the remark this answers is not in the document]
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:237]: It is still called if you disable the new code using the config property.

If we're confident enough in the new approach to not need a flag I can take it out though.

**Comment by alexjo2144** [comment]
@findepi comments applied, PTAL when you get a chance.

**Comment by alexjo2144** [comment]
A test with secrets would also be great

**Comment by findepi** [comment]
/test-with-secrets sha=f9840979650b7774cbe5383eb8c31e9e4aba8a58

**Comment by findepi** [comment]
failure looks related

**Comment by findepi** [comment]
/test-with-secrets sha=b441b7fc458c58754ab0b8607dd488a537325f7d

**Comment by alexjo2144** [comment]
Failure on the run with secrets was in the databricks delta lake suite, unrelated.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableType.java [hunk 2]
```diff
@@ -22,5 +22,6 @@ public enum TableType
     PARTITIONS,
     FILES,
     PROPERTIES,
-    REFS
+    REFS,
+    MATERIALIZED_VIEW_STORAGE
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableType.java:26]:
```suggestion
MATERIALIZED_VIEW_STORAGE,
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/file/FileMetastoreTableOperations.java [hunk 3]
```diff
@@ -53,25 +56,38 @@ public FileMetastoreTableOperations(
     protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
     {
         Table currentTable = getTable();
+        commitTableUpdate(currentTable, metadata, (table, newMetadataLocation) -> Table.builder(table)
+                .apply(builder -> updateMetastoreTable(builder, metadata, newMetadataLocation, Optional.of(currentMetadataLocation)))
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/file/FileMetastoreTableOperations.java:60]:
```suggestion
                .apply(builder -> builder
                        .setParameter(METADATA_LOCATION_PROP, newMetadataLocation)
                        .setParameter(PREVIOUS_METADATA_LOCATION_PROP, currentMetadataLocation))
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java [hunk 4]
```diff
@@ -176,6 +182,8 @@ public void commit(@Nullable TableMetadata base, TableMetadata metadata)
 
     protected abstract void commitToExistingTable(TableMetadata base, TableMetadata metadata);
 
+    protected abstract void commitMaterializedViewRefresh(TableMetadata base, TableMetadata metadata);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java:185]: `commitMaterializedViewRefresh` -> `commitMaterializedView`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java [hunk 5]
```diff
@@ -90,17 +97,25 @@ protected GlueIcebergTableOperations(
...
-        if (isTrinoView(tableType, parameters) || isTrinoMaterializedView(tableType, parameters)) {
+        if (!isMaterializedViewStorageTable && (isTrinoView(tableType, parameters) || isTrinoMaterializedView(tableType, parameters))) {
             // this is a Hive view or Trino/Presto view, or Trino materialized view, hence not a table
             // TODO table operations should not be constructed for views (remove exception-driven code path)
             throw new TableNotFoundException(getSchemaTableName());
         }
-        if (!isIcebergTable(parameters)) {
+        if (!isMaterializedViewStorageTable && !isIcebergTable(parameters)) {
             throw new UnknownTableTypeException(getSchemaTableName());
         }
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:120]:
```suggestion
        if (!isMaterializedViewStorageTable) {
            if (isTrinoView(tableType, parameters) || isTrinoMaterializedView(tableType, parameters)) {
                // this is a Hive view or Trino/Presto view, or Trino materialized view, hence not a table
                // TODO table operations should not be constructed for views (remove exception-driven code path)
                throw new TableNotFoundException(getSchemaTableName());
            }
            if (!isIcebergTable(parameters)) {
                throw new UnknownTableTypeException(getSchemaTableName());
            }
        }
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 6]
```diff
@@ -221,6 +240,35 @@ protected void deleteTableDirectory(TrinoFileSystem fileSystem, SchemaTableName
...
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:243]: If you pass also `fileSystemFactory` to this method, there is no need to pass anymore this field as a parameter to the class's constructor.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java [hunk 7]
```diff
@@ -206,7 +207,7 @@ public void testRefreshMaterializedView()
 
         assertMetastoreInvocations("REFRESH MATERIALIZED VIEW test_refresh_mview_view",
                 ImmutableMultiset.builder()
-                        .addCopies(GET_TABLE, 6)
+                        .addCopies(GET_TABLE, 2)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java:210]: That's quite an improvement. :+1: 

Maybe it is worth adding it in the description of the PR

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 8]
```diff
@@ -341,6 +343,19 @@ public double getMinimumAssignedSplitWeight()
         return minimumAssignedSplitWeight;
     }
 
+    public boolean isHideMaterializedViewStorageTableEnabled()
+    {
+        return hideMaterializedViewStorageTableEnabled;
+    }
+
+    @Config("iceberg.materialized-views.hide-storage-tables")
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:351]: Please add a test using `iceberg.materialized-views.hide-storage-tables=false`.
We need to ensure that Trino can still read/refresh this kind of materialized views.

**Comment by findepi** [comment]
(rebased)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 9]
```diff
@@ -264,6 +283,35 @@ protected void deleteTableDirectory(TrinoFileSystem fileSystem, SchemaTableName
...
+    {
+        String storageTableName = IcebergTableName.tableNameWithType(viewName.getTableName(), MATERIALIZED_VIEW_STORAGE);
+        Map<String, Object> storageTableProperties = new HashMap<>(definition.getProperties());
+        storageTableProperties.putIfAbsent(FILE_FORMAT_PROPERTY, DEFAULT_FILE_FORMAT_DEFAULT);
+
+        String storageSchema = getStorageSchema(definition.getProperties()).orElse(viewName.getSchemaName());
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:292]: Let's skip the call to `getStorageSchema`. It would be good to deprecate the STORAGE_SCHEMA mv property in favor of the new approach.

The STORAGE_SCHEMA was introduced basically to hide those MV storage tables, which we're doing now better.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 10]
```diff
@@ -264,6 +284,37 @@ protected void deleteTableDirectory(TrinoFileSystem fileSystem, SchemaTableName
...
+        String storageTableName = tableNameWithType(viewName.getTableName(), MATERIALIZED_VIEW_STORAGE);
+        Map<String, Object> storageTableProperties = new HashMap<>(definition.getProperties());
+        storageTableProperties.putIfAbsent(FILE_FORMAT_PROPERTY, DEFAULT_FILE_FORMAT_DEFAULT);
+
+        if (getStorageSchema(definition.getProperties()).isPresent()) {
+            throw new TrinoException(NOT_SUPPORTED, "Materialized view property '%s' is not supported when hiding materialized view storage tables is enabled".formatted(STORAGE_SCHEMA));
+        }
+        SchemaTableName storageTable = new SchemaTableName(viewName.getSchemaName(), storageTableName);
+        String tableLocation = defaultTableLocation(session, storageTable);
+        storageTableProperties.put(LOCATION_PROPERTY, tableLocation);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:298]: we allow `location` for CMV, let's honor it

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 11]
```diff
@@ -1195,17 +1225,19 @@ public void updateMaterializedViewColumnComment(ConnectorSession session, Schema
...
                 definition.getProperties());
 
-        updateMaterializedView(session, viewName, newDefinition);
+        updateMaterializedView(viewName, newDefinition);
     }
 
-    private void updateMaterializedView(ConnectorSession session, SchemaTableName viewName, ConnectorMaterializedViewDefinition newDefinition)
+    private void updateMaterializedView(SchemaTableName viewName, ConnectorMaterializedViewDefinition newDefinition)
     {
-        TableInput materializedViewTableInput = getMaterializedViewTableInput(
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:1203]: Why this change?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:1203]: The new code copies the existing MV's table parameters while the old code blows them away. This is important since the paramaters include the storage table metadata path.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:1203]: if so, we could still use `getMaterializedViewTableInput`, just provide `getTableParameters(table)` as parameters, instead of `createMaterializedViewProperties`, right?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:1203]: Sure, that should work.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:1203]: like

```java
com.amazonaws.services.glue.model.Table table = getTable(viewName, false);
TableInput materializedViewTableInput = getMaterializedViewTableInput(
        viewName.getTableName(),
        encodeMaterializedViewData(fromConnectorMaterializedViewDefinition(newDefinition)),
        table.getOwner(),
        getTableParameters(table));
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 12]
```diff
@@ -6701,30 +6701,6 @@ public void testSnapshotSummariesHaveTrinoQueryIdFormatV2()
                 "WHEN MATCHED THEN UPDATE SET b = t.b * 50", tableName, sourceTableName)));
     }
 
-    @Test
-    public void testMaterializedViewSnapshotSummariesHaveTrinoQueryId()
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:6705]: Why this change?

**Comment by findepi** [comment]
/test-with-secrets sha=28892f73b45825172bf0fa552a832753466a602e

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java (hunk 8 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java]: talked offline that this can be a follow-up.

**Comment by mosabua** [comment]
Modified the release notes entry and added.

**Comment by wobrycki** [comment]
Hi @alexjo2144, 
are the materialized view storage tables still accessible as iceberg tables as it was before?
Previously we had a case, when storage table (with name starting from st_) was referenced from spark, connected to the same iceberg configuration as trino. 
For some reason we choose Spark for certain operations on iceberg tables and we would like to continue doing it.

Now, system.metadata.materiazed_views is saying the storage table name is view_name$materialized_view_storage and spark ends with error: `pyspark.sql.utils.AnalysisException: Table or view not found` when trying to access the storage table.
